tests/test_prolongation.cpp

```diff
@@ -0,0 +1,297 @@
+
+    gmgpolar test_p;
+
+    create_grid(test_p);
+
+    level& p_level = *(test_p.v_level[0]);
+    int ctheta_int = test_p.v_level[1]->ntheta_int;
+
+    p_level.m  = test_p.v_level[0]->nr * test_p.v_level[0]->ntheta;
+    p_level.mc = test_p.v_level[1]->nr * test_p.v_level[1]->ntheta;
+
+    std::vector<double> u_test(p_level.mc);
+    for (int z = 0; z < p_level.mc; z++) {
+        u_test[z] = z; //arbitrary grid-function to test the prolongation
+    }
+    std::vector<double> sol = p_level.apply_prolongation_ex(u_test);
+
+    for (int j = 0; j < p_level.nr_int + 1; j++) {
+        for (int i = 0; i < p_level.ntheta_int; i++) {
+            if (j % 2 == 0 && i % 2 == 0) { //coarse node injection
+                EXPECT_EQ(u_test[(j / 2) * ctheta_int + (i / 2)], sol[j * p_level.ntheta_int + i])
+                    << "coarse node injection is failing";
+            }
+            else if (i % 2 != 0 && j % 2 == 0) { // theta_i is fine node. upper edge of the triangle
+
+                int i1 = (j / 2) * ctheta_int + (i - 1) / 2;
+                int i2 = (i < p_level.ntheta_int - 1) ? i1 + 1 : (j / 2) * ctheta_int; //cyclic coordinates
+
+                double val = 0.5 * (u_test[i1] + u_test[i2]);
+
+                EXPECT_NEAR(val, sol[j * p_level.ntheta_int + i], 1e-8)
+                    << "Extrapolated Prolongation fails for Index (r,theta): (" + std::to_string(j) + "," +
+                           std::to_string(i) + ")";
+            }
+            else if (i % 2 == 0) { // r_j fine node, theta_i coarse. lower edge of the triangle
+
+                double v1 = u_test[(j - 1) / 2 * ctheta_int + (i / 2)];
+                double v2 = u_test[(j + 1) / 2 * ctheta_int + (i / 2)];
+
+                double val = 0.5 * (v1 + v2);
+
+                EXPECT_NEAR(val, sol[j * p_level.ntheta_int + i], 1e-8)
+                    << "Extrapolated Prolongation fails for Index (r,theta): (" + std::to_string(j) + "," +
+                           std::to_string(i) + ")";
+            }
+            else // both are fine nodes
+            {
+
+                /*in the triangulation we now consider that the fine node is on the hypothenuse of the triangle.
+                 The corresponding coarse nodes are located on the triangles nodes that span the hypothenuse */
+
+                double top_left;
+                double bottom_right;
+
+                if (i < p_level.ntheta_int - 1) {
+                    top_left     = u_test[((j - 1) / 2) * ctheta_int + (i + 1) / 2];
+                    bottom_right = u_test[((j + 1) / 2) * ctheta_int + (i - 1) / 2];
+                }
+                else {
+                    top_left     = u_test[((j - 1) / 2) * ctheta_int];
+                    bottom_right = u_test[((j + 1) / 2) * ctheta_int + (i - 1) / 2];
+                }
+
+                double val = 0.5 * (top_left + bottom_right);
+
+                EXPECT_NEAR(val, sol[j * p_level.ntheta_int + i], 1e-8)
```

Comment:
```suggestion
                EXPECT_NEAR(val, sol[j * p_level.ntheta_int + i], 1e-12)
```
Can we make this succeed with 1e-12 too?

Reply:
yes! sadly for the bilinear restriction (different file) the rounding error gets too large. The test there works fine for 1e-10